File: transektcount/src/main/java/com/wmstein/transektcount/CountOptionsActivity.java
package com.wmstein.transektcount;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.Toast;

import com.wmstein.transektcount.database.Alert;
import com.wmstein.transektcount.database.AlertDataSource;
import com.wmstein.transektcount.database.Count;
import com.wmstein.transektcount.database.CountDataSource;
import com.wmstein.transektcount.widgets.AddAlertWidget;
import com.wmstein.transektcount.widgets.AlertCreateWidget;
import com.wmstein.transektcount.widgets.EditNotesWidget;
import com.wmstein.transektcount.widgets.OptionsWidget;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.NavUtils;

/************************************************************
 * Edit options for counting species
 * uses optionsWidget.java and widget_options.xml
 * Supplemented with functions for transect external counter
 * Based on CountOptionsActivity.java by milo on 05/05/2014.
 * Adapted and changed by wmstein since 2016-02-18,
 * last edited on 2023-05-08
 */
public class CountOptionsActivity extends AppCompatActivity implements SharedPreferences.OnSharedPreferenceChangeListener
{
    private static final String TAG = "transektcountCntOptAct";
    @SuppressLint("StaticFieldLeak")
    private static TransektCountApplication transektCount;
    
    SharedPreferences prefs;

    private Count count;
    private int count_id;
    private CountDataSource countDataSource;
    private AlertDataSource alertDataSource;
    private View markedForDelete;
    private int deleteAnAlert;
    private int section_id;

    private Bitmap bMap;
    private BitmapDrawable bg;

    LinearLayout static_widget_area;
    LinearLayout dynamic_widget_area;
    OptionsWidget curr_val_widget;
    EditNotesWidget enw;
    AddAlertWidget aa_widget;
    private boolean brightPref;

    ArrayList<AlertCreateWidget> savedAlerts;

    @SuppressLint("SourceLockedOrientationActivity")
    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_count_options);

        transektCount = (TransektCountApplication) getApplication();
        prefs = com.wmstein.transektcount.TransektCountApplication.getPrefs();
        prefs.registerOnSharedPreferenceChangeListener(this);
        brightPref = prefs.getBoolean("pref_bright", true);

        // Set full brightness of screen
        if (brightPref)
        {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
            WindowManager.LayoutParams params = getWindow().getAttributes();
            params.screenBrightness = 1.0f;
            getWindow().setAttributes(params);
        }

        ScrollView counting_screen = findViewById(R.id.count_options);
        bMap = transektCount.decodeBitmap(R.drawable.kbackground, transektCount.width, transektCount.height);
        bg = new BitmapDrawable(counting_screen.getResources(), bMap);
        counting_screen.setBackground(bg);

        static_widget_area = findViewById(R.id.static_widget_area);
        dynamic_widget_area = findViewById(R.id.dynamic_widget_area);

        Bundle extras = getIntent().getExtras();
        if (extras != null)
        {
            count_id = extras.getInt("count_id");
            section_id = extras.getInt("section_id");
            //spinnerPosition = extras.getInt("itemposition");
        }

        savedAlerts = new ArrayList<>();
        if (savedInstanceState != null)
        {
            if (savedInstanceState.getSerializable("savedAlerts") != null)
            {
                //noinspection unchecked
                savedAlerts = (ArrayList<AlertCreateWidget>) savedInstanceState.getSerializable("savedAlerts");
            }
        }
    }

    @Override
    protected void onResume()
    {
        super.onResume();

        // build the count options screen
        // clear any existing views
        static_widget_area.removeAllViews();
        dynamic_widget_area.removeAllViews();

        // get the data sources
        countDataSource = new CountDataSource(this);
        countDataSource.open();
        alertDataSource = new AlertDataSource(this);
        alertDataSource.open();

        count = countDataSource.getCountById(count_id);

        try
        {
            Objects.requireNonNull(getSupportActionBar()).setTitle(count.name);
        } catch (NullPointerException e)
        {
            if (MyDebug.LOG)
                Log.e(TAG, "Problem setting title bar: " + e);
        }

        List<Alert> alerts = alertDataSource.getAllAlertsForCount(count_id);

        // setup the static widgets in the following order
        // 1. Current count value (internal counter)
        // 2. Current counta value (external counter)
        // 3. Alert add/remove
        curr_val_widget = new OptionsWidget(this, null);
        curr_val_widget.setInstructionsf1i(String.format(getString(R.string.editCountValuef1i), count.count_f1i));
        curr_val_widget.setInstructionsf2i(String.format(getString(R.string.editCountValuef2i), count.count_f2i));
        curr_val_widget.setInstructionsf3i(String.format(getString(R.string.editCountValuef3i), count.count_f3i));
        curr_val_widget.setInstructionspi(String.format(getString(R.string.editCountValuepi), count.count_pi));
        curr_val_widget.setInstructionsli(String.format(getString(R.string.editCountValueli), count.count_li));
        curr_val_widget.setInstructionsei(String.format(getString(R.string.editCountValueei), count.count_ei));
        curr_val_widget.setInstructionsf1e(String.format(getString(R.string.editCountValuef1e), count.count_f1e));
        curr_val_widget.setInstructionsf2e(String.format(getString(R.string.editCountValuef2e), count.count_f2e));
        curr_val_widget.setInstructionsf3e(String.format(getString(R.string.editCountValuef3e), count.count_f3e));
        curr_val_widget.setInstructionspe(String.format(getString(R.string.editCountValuepe), count.count_pe));
        curr_val_widget.setInstructionsle(String.format(getString(R.string.editCountValuele), count.count_le));
        curr_val_widget.setInstructionsee(String.format(getString(R.string.editCountValueee), count.count_ee));
        
        curr_val_widget.setParameterValuef1i(count.count_f1i);
        curr_val_widget.setParameterValuef2i(count.count_f2i);
        curr_val_widget.setParameterValuef3i(count.count_f3i);
        curr_val_widget.setParameterValuepi(count.count_pi);
        curr_val_widget.setParameterValueli(count.count_li);
        curr_val_widget.setParameterValueei(count.count_ei);
        curr_val_widget.setParameterValuef1e(count.count_f1e);
        curr_val_widget.setParameterValuef2e(count.count_f2e);
        curr_val_widget.setParameterValuef3e(count.count_f3e);
        curr_val_widget.setParameterValuepe(count.count_pe);
        curr_val_widget.setParameterValuele(count.count_le);
        curr_val_widget.setParameterValueee(count.count_ee);
        
        static_widget_area.addView(curr_val_widget);

        enw = new EditNotesWidget(this, null);
        enw.setSectionNotes(count.notes);
        enw.setWidgetNotes(getString(R.string.notesSpecies));
        enw.setHint(getString(R.string.notesHint));

        static_widget_area.addView(enw);

        aa_widget = new AddAlertWidget(this, null);
        static_widget_area.addView(aa_widget);

    /*
     * There should be a method to add all counts in order to re-draw when one is deleted.
     */
        for (Alert alert : alerts)
        {
            AlertCreateWidget acw = new AlertCreateWidget(this, null);
            acw.setAlertName(alert.alert_text);
            acw.setAlertValue(alert.alert);
            acw.setAlertId(alert.id);
            dynamic_widget_area.addView(acw);
        }

    /*
     * Add saved alert create widgets
     */
        for (AlertCreateWidget acw : savedAlerts)
        {
            dynamic_widget_area.addView(acw);
        }
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState)
    {
    /*
     * Before these widgets can be serialised they must be removed from their parent, or else
     * trying to add them to a new parent causes a crash because they've already got one.
     */
        super.onSaveInstanceState(outState);
        for (AlertCreateWidget acw : savedAlerts)
        {
            ((ViewGroup) acw.getParent()).removeView(acw);
        }
        outState.putSerializable("savedAlerts", savedAlerts);
    }

    @Override
    protected void onPause()
    {
        super.onPause();

        // finally, close the database
        countDataSource.close();
        alertDataSource.close();

    }

    public void saveAndExit(View view)
    {
        saveData();
        savedAlerts.clear();
        super.finish();
    }

    @SuppressLint("LongLogTag")
    public void saveData()
    {
        // don't crash if the user hasn't filled things in...
        // Toast here, as snackbar doesn't show up
        Toast.makeText(CountOptionsActivity.this, getString(R.string.sectSaving) + " " + count.name + "!", Toast.LENGTH_SHORT).show();
        count.count_f1i = curr_val_widget.getParameterValuef1i();
        count.count_f2i = curr_val_widget.getParameterValuef2i();
        count.count_f3i = curr_val_widget.getParameterValuef3i();
        count.count_pi = curr_val_widget.getParameterValuepi();
        count.count_li = curr_val_widget.getParameterValueli();
        count.count_ei = curr_val_widget.getParameterValueei();
        count.count_f1e = curr_val_widget.getParameterValuef1e();
        count.count_f2e = curr_val_widget.getParameterValuef2e();
        count.count_f3e = curr_val_widget.getParameterValuef3e();
        count.count_pe = curr_val_widget.getParameterValuepe();
        count.count_le = curr_val_widget.getParameterValuele();
        count.count_ee = curr_val_widget.getParameterValueee();
        count.notes = enw.getSectionNotes();

        countDataSource.saveCount(count);

    /*
     * Get all the alerts from the dynamic_widget_area and save each one.
     * If it has an id value set to anything higher than 0 then it should be an update, if it is 0
     * then it's a new alert and should be created instead.
     */
        int childcount = dynamic_widget_area.getChildCount();
        for (int i = 0; i < childcount; i++)
        {
            AlertCreateWidget acw = (AlertCreateWidget) dynamic_widget_area.getChildAt(i);
            if (isNotEmpty(acw.getAlertName()))
            {
                // save or create
                if (acw.getAlertId() == 0)
                {
                    alertDataSource.createAlert(count_id, acw.getAlertValue(), acw.getAlertName());
                }
                else
                {
                    alertDataSource.saveAlert(acw.getAlertId(), acw.getAlertValue(), acw.getAlertName());
                }
            }
            else
            {
                if (MyDebug.LOG)
                    Log.d(TAG, "Failed to save alert: " + acw.getAlertId());
            }
        }
    }

    // Scroll to end of view, by wmstein
    public void ScrollToEndOfView(View scrlV)
    {
        int scroll_amount = scrlV.getBottom();
        int scrollY = scroll_amount;
        boolean pageend = false;
        while (!pageend)
        {
            scrlV.scrollTo(0, scroll_amount);           //scroll
            scroll_amount = scroll_amount + scroll_amount; //increase scroll_amount
            scrollY = scrollY + scrlV.getScrollY();        //scroll position of 1. row
            if (scroll_amount > scrollY)
            {
                pageend = true;
            }
        }
    }

    // Add alert to species counter
    public void addAnAlert(View view)
    {
        AlertCreateWidget acw = new AlertCreateWidget(this, null);
        savedAlerts.add(acw);
        // Scroll to end of view
        View scrollV = findViewById(R.id.count_options);
        ScrollToEndOfView(scrollV);
        acw.requestFocus();
        dynamic_widget_area.addView(acw);
    }

    // Delete alert from species counter and its widget from the view
    public void deleteWidget(View view)
    {
    /*
     * These global variables keep a track of the view containing an alert to be deleted and also the id
     * of the alert itself, to make sure that they're available inside the code for the alert dialog by
     * which they will be deleted.
     */
        AlertDialog.Builder areYouSure;
        markedForDelete = view;
        deleteAnAlert = (Integer) view.getTag();
        if (deleteAnAlert == 0)
        {
            //Log.i(TAG, "(1) View tag was " + String.valueOf(deleteAnAlert));
            // the actual AlertCreateWidget is two levels up from the button in which it is embedded
            dynamic_widget_area.removeView((AlertCreateWidget) view.getParent().getParent());
        }
        else
        {
            //Log.i(TAG, "(2) View tag was " + String.valueOf(deleteAnAlert));
            // before removing this widget it is necessary to do the following:
            // (1) Check the user is sure they want to delete it and, if so...
            // (2) Delete the associated alert from the database.
            areYouSure = new AlertDialog.Builder(this);
            areYouSure.setTitle(getString(R.string.deleteAlert));
            areYouSure.setMessage(getString(R.string.reallyDeleteAlert));
            areYouSure.setPositiveButton(R.string.yesDeleteIt, (dialog, whichButton) -> {
                // go ahead for the delete
                try
                {
                    alertDataSource.deleteAlertById(deleteAnAlert);
                    dynamic_widget_area.removeView((AlertCreateWidget) markedForDelete.getParent().getParent());
                } catch (Exception e)
                {
                    if (MyDebug.LOG)
                        Log.e(TAG, "Failed to delete a widget: " + e);
                }
            });
            areYouSure.setNegativeButton(R.string.cancel, (dialog, whichButton) -> {
                // Cancelled.
            });
            areYouSure.show();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu)
    {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.count_options, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();
        if (id == android.R.id.home)
        {
            Intent intent = NavUtils.getParentActivityIntent(this);
            assert intent != null;
            intent.putExtra("section_id", section_id);
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
            NavUtils.navigateUpTo(this, intent);
        }
        else if (id == R.id.menuSaveExit)
        {
            saveData();
            super.finish();
        }
        return super.onOptionsItemSelected(item);
    }

    @SuppressLint("SourceLockedOrientationActivity")
    public void onSharedPreferenceChanged(SharedPreferences prefs, String key)
    {
        ScrollView counting_screen = findViewById(R.id.count_options);
        counting_screen.setBackground(null);
        brightPref = prefs.getBoolean("pref_bright", true);

        bMap = transektCount.decodeBitmap(R.drawable.kbackground, transektCount.width, transektCount.height);
        bg = new BitmapDrawable(counting_screen.getResources(), bMap);
        counting_screen.setBackground(bg);
    }

    /**
     * Following functions are taken from the Apache commons-lang3-3.4 library
     * licensed under Apache License Version 2.0, January 2004
     <p> 
     * Checks if a CharSequence is not empty ("") and not null.
     <p> 
     * isNotEmpty(null)      = false
     * isNotEmpty("")        = false
     * isNotEmpty(" ")       = true
     * isNotEmpty("bob")     = true
     * isNotEmpty("  bob  ") = true
     *
     * @param cs the CharSequence to check, may be null
     * @return {@code true} if the CharSequence is not empty and not null
     */
    public static boolean isNotEmpty(final CharSequence cs)
    {
        return !isEmpty(cs);
    }

    /**
     * Checks if a CharSequence is empty ("") or null.
     <p> 
     * isEmpty(null)      = true
     * isEmpty("")        = true
     * isEmpty(" ")       = false
     * isEmpty("bob")     = false
     * isEmpty("  bob  ") = false
     *
     * @param cs the CharSequence to check, may be null
     * @return {@code true} if the CharSequence is empty or null
     */
    public static boolean isEmpty(final CharSequence cs)
    {
        return cs == null || cs.length() == 0;
    }

}

File: transektcount/src/main/java/com/wmstein/transektcount/CountOptionsActivity.kt
@file:Suppress("USELESS_CAST")

package com.wmstein.transektcount

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.ViewGroup.MarginLayoutParams
import android.view.WindowManager
import android.widget.LinearLayout
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NavUtils
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updateLayoutParams
import com.wmstein.transektcount.Utils.fromHtml
import com.wmstein.transektcount.database.Count
import com.wmstein.transektcount.database.CountDataSource
import com.wmstein.transektcount.database.SectionDataSource
import com.wmstein.transektcount.widgets.EditSectionListNotesWidget
import com.wmstein.transektcount.widgets.CountOptionsWidget
import com.wmstein.transektcount.widgets.CountOptionsLhWidget

/************************************************************
 * CountOptionsActivity
 * Edit options for counting species
 * uses optionsWidget.kt, optionsWidgetExt.kt and widget_options.xml
 * Supplemented with functions for transect external counter
 * Based on CountOptionsActivity.java by milo on 05/05/2014.
 * Adapted and changed by wmstein since 2016-02-18,
 * last edited in Java on 2023-05-08,
 * converted to Kotlin on 2023-07-17,
 * last edited on 2026-02-22
 */
class CountOptionsActivity : AppCompatActivity() {
    private var count: Count? = null
    private var countId = 0
    private var countDataSource: CountDataSource? = null
    private var sectionDataSource: SectionDataSource? = null
    private var sectionId = 0
    private var sectionName = ""
    private var staticWidgetArea: LinearLayout? = null
    private var dynamicWidgetArea: LinearLayout? = null
    private var optWidget: CountOptionsWidget? = null
    private var optWidgetLh: CountOptionsLhWidget? = null
    private var enw: EditSectionListNotesWidget? = null

    private var sf1i = 0
    private var sf2i = 0
    private var sf3i = 0
    private var spi = 0
    private var sli = 0
    private var sei = 0
    private var sf1e = 0
    private var sf2e = 0
    private var sf3e = 0
    private var spe = 0
    private var sle = 0
    private var see = 0

    // Criteria on setting or removing the date for section:
    //  countsInit = false, countsCurrent = false -> delete date
    //  countsInit = false, countsCurrent = true -> save date
    //  countsInit = true, countsCurrent = false -> delete date
    //  countsInit = true, countsCurrent = true -> no change
    private var countsInit = false
    private var countsCurrent = false

    // Preferences
    private var prefs = TransektCountApplication.getPrefs()
    private var brightPref = false
    private var lhandPref = false
    private var awakePref = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (IsRunningOnEmulator.DLOG || BuildConfig.DEBUG)
            Log.i(TAG, "85, onCreate")

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.VANILLA_ICE_CREAM) // SDK 35+
        {
            enableEdgeToEdge()
        }

        setContentView(R.layout.activity_count_options)

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.count_options)) { v, windowInsets ->
            val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())
            // Apply the insets as a margin to the view. This solution sets
            // only the bottom, left, and right dimensions, but you can apply whichever
            // insets are appropriate to your layout. You can also update the view padding
            // if that's more appropriate.
            v.updateLayoutParams<MarginLayoutParams> {
                topMargin = insets.top
                leftMargin = insets.left
                bottomMargin = insets.bottom
                rightMargin = insets.right
            }

            // Return CONSUMED if you don't want the window insets to keep passing
            // down to descendant views.
            WindowInsetsCompat.CONSUMED
        }

        val extras = intent.extras
        if (extras != null) {
            countId = extras.getInt("count_id")
            sectionId = extras.getInt("section_id")
            sectionName = extras.getString("section_name").toString()
        }

        staticWidgetArea = findViewById(R.id.static_widget_area)
        dynamicWidgetArea = findViewById(R.id.dynamic_widget_area)

        countDataSource = CountDataSource(this)
        sectionDataSource = SectionDataSource(this)

        // New onBackPressed logic
        val callback = object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                val intent = NavUtils.getParentActivityIntent(this@CountOptionsActivity)!!
                intent.putExtra("section_id", sectionId)
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                this@CountOptionsActivity.navigateUpTo(intent)
            }
        }
        onBackPressedDispatcher.addCallback(this, callback)
    }
    // End of onCreate()

    override fun onResume() {
        super.onResume()

        if (IsRunningOnEmulator.DLOG || BuildConfig.DEBUG)
            Log.i(TAG, "142, onResume")

        brightPref = prefs.getBoolean("pref_bright", true)
        awakePref = prefs.getBoolean("pref_awake", true)
        lhandPref = prefs.getBoolean("pref_left_hand", false)

        // Set full brightness of screen
        if (brightPref) {
            val params = window.attributes
            params.screenBrightness = 1.0f
            window.attributes = params
        }

        if (awakePref)
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        // Clear any existing views
        staticWidgetArea!!.removeAllViews()
        dynamicWidgetArea!!.removeAllViews()

        // Get the data sources
        countDataSource!!.open()
        sectionDataSource!!.open()
        count = countDataSource!!.getCountById(countId)

        supportActionBar!!.title = count!!.name
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        // Set up the static widgets in the following order
        //  1. Current count values (internal counters)
        //  2. Current count values (external counters)
        if (lhandPref) {
            optWidgetLh = CountOptionsLhWidget(this, null)
            sf1i = optWidgetLh!!.parameterValuef1i
            sf2i = optWidgetLh!!.parameterValuef2i
            sf3i = optWidgetLh!!.parameterValuef3i
            spi = optWidgetLh!!.parameterValuepi
            sli = optWidgetLh!!.parameterValueli
            sei = optWidgetLh!!.parameterValueei

            sf1e = optWidgetLh!!.parameterValuef1e
            sf2e = optWidgetLh!!.parameterValuef2e
            sf3e = optWidgetLh!!.parameterValuef3e
            spe = optWidgetLh!!.parameterValuepe
            sle = optWidgetLh!!.parameterValuele
            see = optWidgetLh!!.parameterValueee
        } else {
            optWidget = CountOptionsWidget(this, null)
            sf1i = optWidget!!.parameterValuef1i
            sf2i = optWidget!!.parameterValuef2i
            sf3i = optWidget!!.parameterValuef3i
            spi = optWidget!!.parameterValuepi
            sli = optWidget!!.parameterValueli
            sei = optWidget!!.parameterValueei

            sf1e = optWidget!!.parameterValuef1e
            sf2e = optWidget!!.parameterValuef2e
            sf3e = optWidget!!.parameterValuef3e
            spe = optWidget!!.parameterValuepe
            sle = optWidget!!.parameterValuele
            see = optWidget!!.parameterValueee
        }
        if (sf1i + sf2i + sf3i + spi + sli + sei + sf1e + sf2e + sf3e + spe + sle + see > 0)
            countsInit = true // has initial counts
        else
            countsInit = false // no count so far

        if (lhandPref) {
            optWidgetLh!!.setInstructionsf1i(
                String.format(
                    getString(R.string.editCountValuef1i),
                    count!!.count_f1i
                )
            )
            optWidgetLh!!.setInstructionsf2i(
                String.format(
                    getString(R.string.editCountValuef2i),
                    count!!.count_f2i
                )
            )
            optWidgetLh!!.setInstructionsf3i(
                String.format(
                    getString(R.string.editCountValuef3i),
                    count!!.count_f3i
                )
            )
            optWidgetLh!!.setInstructionspi(
                String.format(
                    getString(R.string.editCountValuepi),
                    count!!.count_pi
                )
            )
            optWidgetLh!!.setInstructionsli(
                String.format(
                    getString(R.string.editCountValueli),
                    count!!.count_li
                )
            )
            optWidgetLh!!.setInstructionsei(
                String.format(
                    getString(R.string.editCountValueei),
                    count!!.count_ei
                )
            )
            optWidgetLh!!.setInstructionsf1e(
                String.format(
                    getString(R.string.editCountValuef1e),
                    count!!.count_f1e
                )
            )
            optWidgetLh!!.setInstructionsf2e(
                String.format(
                    getString(R.string.editCountValuef2e),
                    count!!.count_f2e
                )
            )
            optWidgetLh!!.setInstructionsf3e(
                String.format(
                    getString(R.string.editCountValuef3e),
                    count!!.count_f3e
                )
            )
            optWidgetLh!!.setInstructionspe(
                String.format(
                    getString(R.string.editCountValuepe),
                    count!!.count_pe
                )
            )
            optWidgetLh!!.setInstructionsle(
                String.format(
                    getString(R.string.editCountValuele),
                    count!!.count_le
                )
            )
            optWidgetLh!!.setInstructionsee(
                String.format(
                    getString(R.string.editCountValueee),
                    count!!.count_ee
                )
            )
            optWidgetLh!!.parameterValuef1i = count!!.count_f1i
            optWidgetLh!!.parameterValuef2i = count!!.count_f2i
            optWidgetLh!!.parameterValuef3i = count!!.count_f3i
            optWidgetLh!!.parameterValuepi = count!!.count_pi
            optWidgetLh!!.parameterValueli = count!!.count_li
            optWidgetLh!!.parameterValueei = count!!.count_ei
            optWidgetLh!!.parameterValuef1e = count!!.count_f1e
            optWidgetLh!!.parameterValuef2e = count!!.count_f2e
            optWidgetLh!!.parameterValuef3e = count!!.count_f3e
            optWidgetLh!!.parameterValuepe = count!!.count_pe
            optWidgetLh!!.parameterValuele = count!!.count_le
            optWidgetLh!!.parameterValueee = count!!.count_ee
            staticWidgetArea!!.addView(optWidgetLh)
        } else {
            optWidget!!.setInstructionsf1i(
                String.format(
                    getString(R.string.editCountValuef1i),
                    count!!.count_f1i
                )
            )
            optWidget!!.setInstructionsf2i(
                String.format(
                    getString(R.string.editCountValuef2i),
                    count!!.count_f2i
                )
            )
            optWidget!!.setInstructionsf3i(
                String.format(
                    getString(R.string.editCountValuef3i),
                    count!!.count_f3i
                )
            )
            optWidget!!.setInstructionspi(
                String.format(
                    getString(R.string.editCountValuepi),
                    count!!.count_pi
                )
            )
            optWidget!!.setInstructionsli(
                String.format(
                    getString(R.string.editCountValueli),
                    count!!.count_li
                )
            )
            optWidget!!.setInstructionsei(
                String.format(
                    getString(R.string.editCountValueei),
                    count!!.count_ei
                )
            )
            optWidget!!.setInstructionsf1e(
                String.format(
                    getString(R.string.editCountValuef1e),
                    count!!.count_f1e
                )
            )
            optWidget!!.setInstructionsf2e(
                String.format(
                    getString(R.string.editCountValuef2e),
                    count!!.count_f2e
                )
            )
            optWidget!!.setInstructionsf3e(
                String.format(
                    getString(R.string.editCountValuef3e),
                    count!!.count_f3e
                )
            )
            optWidget!!.setInstructionspe(
                String.format(
                    getString(R.string.editCountValuepe),
                    count!!.count_pe
                )
            )
            optWidget!!.setInstructionsle(
                String.format(
                    getString(R.string.editCountValuele),
                    count!!.count_le
                )
            )
            optWidget!!.setInstructionsee(
                String.format(
                    getString(R.string.editCountValueee),
                    count!!.count_ee
                )
            )
            optWidget!!.parameterValuef1i = count!!.count_f1i
            optWidget!!.parameterValuef2i = count!!.count_f2i
            optWidget!!.parameterValuef3i = count!!.count_f3i
            optWidget!!.parameterValuepi = count!!.count_pi
            optWidget!!.parameterValueli = count!!.count_li
            optWidget!!.parameterValueei = count!!.count_ei
            optWidget!!.parameterValuef1e = count!!.count_f1e
            optWidget!!.parameterValuef2e = count!!.count_f2e
            optWidget!!.parameterValuef3e = count!!.count_f3e
            optWidget!!.parameterValuepe = count!!.count_pe
            optWidget!!.parameterValuele = count!!.count_le
            optWidget!!.parameterValueee = count!!.count_ee
            staticWidgetArea!!.addView(optWidget)
        }

        enw = EditSectionListNotesWidget(this, null)
        enw!!.sNotes = count!!.notes
        enw!!.setWidgetNotes(getString(R.string.notesSpecies, sectionName))
        enw!!.setHint(getString(R.string.notesHint))
        staticWidgetArea!!.addView(enw)
    }
    // End of onResume()

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.count_options, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here.
        val id = item.itemId
        if (id == android.R.id.home) {
            val intent = NavUtils.getParentActivityIntent(this)!!
            intent.putExtra("section_id", sectionId)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
            this@CountOptionsActivity.navigateUpTo(intent)
            return true
        } else if (id == R.id.menuSaveExit) {
            if (saveData()) {
                val intent = NavUtils.getParentActivityIntent(this)!!
                intent.putExtra("section_id", sectionId)
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                this@CountOptionsActivity.navigateUpTo(intent)
            }
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPause() {
        super.onPause()

        if (IsRunningOnEmulator.DLOG || BuildConfig.DEBUG)
            Log.i(TAG, "422, onPause")

        countDataSource!!.close()
        sectionDataSource!!.close()

        if (awakePref) {
            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }

        staticWidgetArea!!.clearFocus()
        staticWidgetArea!!.removeAllViews()
        dynamicWidgetArea!!.clearFocus()
        dynamicWidgetArea!!.removeAllViews()
    }

    override fun onStop() {
        super.onStop()

        if (IsRunningOnEmulator.DLOG || BuildConfig.DEBUG)
            Log.i(TAG, "441, onStop")

        staticWidgetArea = null
        dynamicWidgetArea = null
    }

    override fun onDestroy() {
        super.onDestroy()

        if (IsRunningOnEmulator.DLOG || BuildConfig.DEBUG)
            Log.i(TAG, "451, onDestroy")
    }

    fun saveData(): Boolean {
        sf1i = optWidget!!.parameterValuef1i
        count!!.count_f1i = sf1i
        sf2i = optWidget!!.parameterValuef2i
        count!!.count_f2i = sf2i
        sf3i = optWidget!!.parameterValuef3i
        count!!.count_f3i = sf3i
        spi = optWidget!!.parameterValuepi
        count!!.count_pi = spi
        sli = optWidget!!.parameterValueli
        count!!.count_li = sli
        sei = optWidget!!.parameterValueei
        count!!.count_ei = sei
        sf1e = optWidget!!.parameterValuef1e
        count!!.count_f1e = sf1e
        sf2e = optWidget!!.parameterValuef2e
        count!!.count_f2e = sf2e
        sf3e = optWidget!!.parameterValuef3e
        count!!.count_f3e = sf3e
        spe = optWidget!!.parameterValuepe
        count!!.count_pe = spe
        sle = optWidget!!.parameterValuele
        count!!.count_le = sle
        see = optWidget!!.parameterValueee
        count!!.count_ee = see
        count!!.notes = enw!!.sNotes

        val mesg = getString(R.string.saving) + " " + count!!.name + "!"
        Toast.makeText(
            this, fromHtml("<font color='#008000'>$mesg</font>"),
            Toast.LENGTH_SHORT
        ).show()

        if (sf1i + sf2i + sf3i + spi + sli + sei + sf1e + sf2e + sf3e + spe + sle + see > 0)
            countsCurrent = true // has count(s)

        countDataSource!!.saveCount(count!!)
        if (!countsInit && countsCurrent) // if no previous count and a new count entered 
            sectionDataSource!!.saveDateSectionOfId(sectionId)

        if (!countsCurrent) // if no count
            sectionDataSource!!.clearDateSectionOfId(sectionId)

        return true
    }

    companion object {
        private const val TAG = "CntOptAct"
    }

}
